include/alpaka/rand/Philox/PhiloxBaseArrayLike.hpp

```diff
@@ -0,0 +1,106 @@
+ */
+
+#pragma once
+
+#include <alpaka/rand/Philox/PhiloxConstants.hpp>
+#include <alpaka/rand/Philox/mulhilo.hpp>
+
+#include <utility>
+
+namespace alpaka
+{
+    namespace rand
+    {
+        namespace engine
+        {
+            /** Base class for Counter and Key types with array-like interface
+             *
+             * Provides the implementation of key bumping, counter advancing, counter skipping, and subsequence
+             * skipping for types that support subscripting by operator [].
+             *
+             * @tparam TCounter Counter type
+             * @tparam TKey Key type
+             * @tparam TParams Philox algorithm parameters \sa PhiloxParams
+             * @tparam TImpl engine type implementation (CRTP)
+             */
+            template<typename TParams, typename TCounter, typename TKey, typename TImpl>
+            class PhiloxBaseArrayLike : public PhiloxConstants<TParams>
+            {
+            protected:
+                using Counter = TCounter;
+                using Key = TKey;
+
+                /** Bump the \a key by the Weyl sequence step parameter
+                 *
+                 * @param key the key to be bumped
+                 * @return the bumped key
+                 */
+                ALPAKA_FN_HOST_ACC auto bumpKey(Key const& key)
+                {
+                    return Key{key[0] + this->WEYL_32_0, key[1] + this->WEYL_32_1};
+                }
+
+                /** Advance the \a counter to the next state
+                 *
+                 * Increments the passed-in \a counter by one with a 128-bit carry.
+                 *
+                 * @param counter reference to the counter which is to be advanced
+                 */
+                ALPAKA_FN_HOST_ACC void advanceCounter(Counter& counter)
+                {
+                    counter[0]++;
+                    /* 128-bit carry */
+                    if(counter[0] == 0)
+                    {
+                        counter[1]++;
+                        if(counter[1] == 0)
+                        {
+                            counter[2]++;
+                            if(counter[2] == 0)
+                            {
+                                counter[3]++;
+                            }
+                        }
+                    }
+                }
+
+                /** Advance the internal state counter by \a offset N-vectors (N = counter size)
+                 *
+                 * Advances the internal value of this->state.counter
+                 *
+                 * @param offset number of N-vectors to skip
+                 */
+                ALPAKA_FN_HOST_ACC void skip4(uint64_t offset)
+                {
+                    Counter& counter = static_cast<TImpl*>(this)->state.counter;
+                    Counter temp = counter;
+                    counter[0] += lo32(offset);
+                    counter[1] += hi32(offset) + (counter[0] < temp[0] ? 1 : 0);
```

Comment:
`lo32` and `hi32` have a bit too short of a name for my taste. Furthermore, I guess this code assumes that `counter[i]` have 32 bits? If yes, place `static_assert` this.

Reply:
I have no issues with renaming them. Any suggestions?

Reply:
There's more assumptions at play here - Counter has 4 elements, Key has 2 elements, and the elements are all 32 bit. I want to provide all four [reference Philox variants](https://www.thesalmons.org/john/random123/releases/latest/docs/group__PhiloxNxW.html) - with 32/64-bit elements, and a 2/4-element counter. (With a customizable number of rounds - 10 for a safe default, 7 for improved performance but less safety margin for those who know what they're doing).

I'll fix this when the other variants are added - that is after I validate the default one.

Reply:
Ok.

Reply:
Actually I remembered this conversation, forgot I didn't want to rename these and renamed them anyway in the current version :-D Now the master function is `multiplyAndSplit64to32` and then there's `high32Bits` and `low32Bits`.